### PR TITLE
LibGfx/GIF: Support loop_count and duration

### PR DESCRIPTION
**Pull request by LucasChollet**

### CODE REVIEW

File: Userland/Libraries/LibGfx/ImageFormats/GIFWriter.cpp [hunk 1]
```diff
@@ -155,6 +188,7 @@ ErrorOr<void> GIFAnimationWriter::add_frame(Bitmap& bitmap, int duration_ms, Int
 
     // Write a Table-Based Image
     BigEndianOutputBitStream bit_stream { MaybeOwned { m_stream } };
+    TRY(write_graphic_control_extension(bit_stream, duration_ms));
```
Comment (nico) [inline Userland/Libraries/LibGfx/ImageFormats/GIFWriter.cpp:191]: nit: commit description says "this is quite useless until we write animations", but we already do

File: Userland/Libraries/LibGfx/ImageFormats/GIFWriter.cpp [hunk 2]
```diff
@@ -193,6 +223,10 @@ ErrorOr<NonnullOwnPtr<AnimationWriter>> GIFWriter::start_encoding_animation(Seek
 
     BigEndianOutputBitStream bit_stream { MaybeOwned<Stream> { stream } };
     TRY(write_logical_descriptor(bit_stream, dimensions));
+
+    // Vendor extension to support looping
+    TRY(write_netscape_extension(bit_stream));
```
Comment (nico) [inline Userland/Libraries/LibGfx/ImageFormats/GIFWriter.cpp:262]: `WebPWriter::start_encoding_animation()` takes a loop_count parameter – do you want to add a parameter too?

File: Tests/LibGfx/TestImageWriter.cpp [hunk 3]
```diff
@@ -113,20 +113,30 @@ TEST_CASE(test_bmp)
...
     // Let's limit the size of the image so every color can fit in a color table of 256 elements.
-    auto bitmap = TRY_OR_FAIL(TRY_OR_FAIL(create_test_rgb_bitmap())->cropped({ 0, 0, 16, 16 }));
+    auto bitmap_1 = TRY_OR_FAIL(TRY_OR_FAIL(create_test_rgb_bitmap())->cropped({ 0, 0, 16, 16 }));
+    auto bitmap_2 = TRY_OR_FAIL(TRY_OR_FAIL(create_test_rgb_bitmap())->cropped({ 16, 16, 16, 16 }));
 
-    auto encoded_bitmap = TRY_OR_FAIL((encode_bitmap<Gfx::GIFWriter>(bitmap)));
-    auto decoder = TRY_OR_FAIL(Gfx::GIFImageDecoderPlugin::create(encoded_bitmap));
+    auto stream_buffer = TRY_OR_FAIL(ByteBuffer::create_uninitialized(10 * 1024));
+    FixedMemoryStream stream { Bytes { stream_buffer } };
+    auto animation_writer = TRY_OR_FAIL(Gfx::GIFWriter::start_encoding_animation(stream, bitmap_1->size()));
```
Comment (nico) [inline Tests/LibGfx/TestImageWriter.cpp:121]: Maybe it's nice to keep the test for GIFImageDecoderPlugin::create()? Consider adding a `test_animated_gif` that does what the RHS does and keep the exisiting `test_gif` as-is.

**Review by nico — COMMENTED** [review]
\o/

File: Userland/Libraries/LibGfx/ImageFormats/GIFWriter.cpp [hunk 4]
```diff
@@ -167,6 +167,36 @@ ErrorOr<void> GIFAnimationWriter::add_frame(Bitmap& bitmap, int duration_ms, Int
     return {};
 }
 
+ErrorOr<void> write_netscape_extension(BigEndianOutputBitStream& stream)
+{
+    // This is a vendor extension, its sole usage is to precise the loop count.
+    // I used this link as a source: http://www6.uniovi.es/gifanim/gifabout.htm
```
[the remark this answers is not in the document]
Comment (LucasChollet) [reply Userland/Libraries/LibGfx/ImageFormats/GIFWriter.cpp:207]: Oops

File: Userland/Libraries/LibGfx/ImageFormats/GIFWriter.cpp (hunk 1 above)
Comment (LucasChollet) [reply Userland/Libraries/LibGfx/ImageFormats/GIFWriter.cpp:191]: Ah, didn't check that, updated!

File: Userland/Libraries/LibGfx/ImageFormats/GIFWriter.cpp (hunk 2 above)
Comment (LucasChollet) [reply Userland/Libraries/LibGfx/ImageFormats/GIFWriter.cpp:262]: can do!

File: Tests/LibGfx/TestImageWriter.cpp (hunk 3 above)
Comment (LucasChollet) [reply Tests/LibGfx/TestImageWriter.cpp]: Good point, I didn't think of that

File: Userland/Libraries/LibGfx/ImageFormats/GIFWriter.cpp [hunk 5]
```diff
@@ -167,6 +167,36 @@ ErrorOr<void> GIFAnimationWriter::add_frame(Bitmap& bitmap, int duration_ms, Int
     return {};
 }
 
+ErrorOr<void> write_netscape_extension(BigEndianOutputBitStream& stream, u16 loop_count)
+{
+    // This is a vendor extension, its sole usage is to precise the loop count.
```
Comment (trflynn89) [inline Userland/Libraries/LibGfx/ImageFormats/GIFWriter.cpp:206]: What does "is to precise the loop count" mean?

File: Userland/Libraries/LibGfx/ImageFormats/GIFWriter.cpp [hunk 6]
```diff
@@ -167,6 +167,36 @@ ErrorOr<void> GIFAnimationWriter::add_frame(Bitmap& bitmap, int duration_ms, Int
...
+    // I used this link as a source: https://web.archive.org/web/19990418091037/http://www6.uniovi.es/gifanim/gifabout.htm
+
+    // Extension Introducer
+    TRY(stream.write_value<u8>(0x21));
+    // Application Extension Label
+    TRY(stream.write_value<u8>(0xFF));
+
+    // Block Size
+    auto constexpr netscape_signature = "NETSCAPE2.0"sv;
```
Comment (trflynn89) [inline Userland/Libraries/LibGfx/ImageFormats/GIFWriter.cpp:215]: (while I'm here, constexpr on the left please)

File: Userland/Libraries/LibGfx/ImageFormats/GIFWriter.cpp (hunk 5 above)
Comment (LucasChollet) [reply Userland/Libraries/LibGfx/ImageFormats/GIFWriter.cpp]: Not much in English it looks like, I replaced it with "provide".

**Pull request merged**